osbuilder: ubuntu: Switch from multistrap to mmdebstrap

skazi0 opened the pull request:
Multistrap requires usrmerge package which was dropped in Ubuntu 24.04 (Noble). Based on details from [0], the rootfs build process was switched to mmdebstrap.

Some additional minor tweaks were needed around chrony as the version from Noble has very strict systemd sandboxing configured and it doesn't work with readonly root by default.

[0] https://lists.debian.org/debian-dpkg/2023/05/msg00080.html

Fixes: #11245

fidencio commented:
@skazi0, I'm wondering here whether ubuntu will switch to `mmdebstrap` instead of `multistrap` in the future, as newer versions of Ubuntu do have `multistrap` working as expected.

We've achieved a similar results by using the following hack: https://bugs.launchpad.net/ubuntu/+source/iproute2/+bug/2085460

Now, the question is ... what's the way to go in the future in your opinion @skazi0?

skazi0 commented:
I don't see any future for `multistrap`. It was orphaned and removed from Debian (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=914282 and https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1093962). The last maintainer of `multistrap` is also the author of `mmdebstrap` which is said to be `multistrap`'s successor. While `multistrap` is still present in current Ubuntu, I don't think they want to take over the ownership and maintenance of this code unless they have some strong internal need to keep it. I see multiple projects switching to `mmdebstrap` already. I would personally do the same.

fidencio commented:
> I don't see any future for `multistrap`. It was orphaned and removed from Debian (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=914282 and https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1093962). The last maintainer of `multistrap` is also the author of `mmdebstrap` which is said to be `multistrap`'s successor. While `multistrap` is still present in current Ubuntu, I don't think they want to take over the ownership and maintenance of this code unless they have some strong internal need to keep it. I see multiple projects switching to `mmdebstrap` already. I would personally do the same.

Amazing info.
Let me test it locally and I will review it Today!

skazi0 commented on tools/osbuilder/rootfs-builder/ubuntu/rootfs_lib.sh:
```diff
@@ -12,35 +12,21 @@ build_dbus() {
-
-	# For simplicity's sake, use multistrap for foreign and native bootstraps.
-	cat > "$multistrap_conf" << EOF
-[General]
-cleanup=true
-aptsources=Ubuntu
-bootstrap=Ubuntu
-
-[Ubuntu]
-source=$REPO_URL
-keyring=ubuntu-keyring
-suite=$OS_VERSION
-packages=$PACKAGES $EXTRA_PKGS
```
There is `--include "$PACKAGES $EXTRA_PKGS"` in the `mmdebstrap` command below.

fidencio replied:
Amazing, I had missed that.

fidencio reviewed (approved):
I've tested this locally building Ubuntu 25.04 (with some small changes only related to the 25.04), and it works like a charm.

Thanks for the contribution, @skazi0!

skazi0 commented:
I've just noticed some problems with `OS_VERSION=focal`.
The `mmdebstrap` version from Focal only supports comma separated package lists and `--mode unshare` didn't like being used in docker's root env.
I switched `--mode` to `auto` which uses `root` mode inside docker and added some Focal specific code to handle commas.
Good news is that the old "dbus hack" doesn't seem to be needed with `mmdebstrap` so I removed it.

zvonkok commented:
@skazi0 It is related to the changes, but the new rootfs do not have the same set of packages as mutlistrap.

zvonkok commented:
libmnl.so. is a dependency if `iptables-save` I wonder if we're missing the iptables packages in the new way of building.

zvonkok commented:
Looking through the logs, I do not see that we install iptables, a mandatory package for the kata-agent, and several tests were cancelled.

skazi0 commented:
> Looking through the logs, I do not see that we install iptables, a mandatory package for the kata-agent, and several tests were cancelled.

Thanks @zvonkok for checking this! Looks like the `apt` variant is not complete enough. I'll try to bump it.

skazi0 commented:
@zvonkok the rootfs build CI passed now. Are the other failures also caused by my changes?

zvonkok commented:
@skazi0 Let's wait for the CI we want all required tests passing.

fidencio commented:
> @zvonkok the rootfs build CI passed now. Are the other failures also caused by my changes?

Nops, we're good! Thanks @skazi0!

stevenhorsman reviewed (approved):
This looks reasonable to me and the required tests are passing. Thanks @skazi0! @fidencio - does this enable us to switch our image bases from jammy (or oracular for the confidential image) to noble now if we'd wish?
